File: app/src/main/java/com/airfreshener/telegram_sms/sms_send_receiver.java
package com.airfreshener.telegram_sms;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.telephony.SmsManager;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.airfreshener.telegram_sms.config.ProxyConfigV2;
import com.airfreshener.telegram_sms.model.RequestMessage;
import com.airfreshener.telegram_sms.utils.LogUtils;
import com.airfreshener.telegram_sms.utils.NetworkUtils;
import com.airfreshener.telegram_sms.utils.ResendUtils;
import com.airfreshener.telegram_sms.utils.SmsUtils;
import com.airfreshener.telegram_sms.value.const_value;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Objects;

import io.paperdb.Paper;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import static android.content.Context.MODE_PRIVATE;

public class sms_send_receiver extends BroadcastReceiver {
    @Override
    public void onReceive(final Context context, @NotNull Intent intent) {
        Paper.init(context);
        final String TAG = "sms_send_receiver";
        Log.d(TAG, "Receive action: " + intent.getAction());
        Bundle extras = intent.getExtras();
        assert extras != null;
        int sub = extras.getInt("sub_id");
        context.getApplicationContext().unregisterReceiver(this);
        SharedPreferences sharedPreferences = context.getSharedPreferences("data", MODE_PRIVATE);
        if (!sharedPreferences.getBoolean("initialized", false)) {
            Log.i(TAG, "Uninitialized, SMS send receiver is deactivated.");
            return;
        }
        String bot_token = sharedPreferences.getString("bot_token", "");
        String chat_id = sharedPreferences.getString("chat_id", "");
        final RequestMessage request_body = new RequestMessage();
        request_body.chat_id = chat_id;
        String request_uri = NetworkUtils.get_url(bot_token, "sendMessage");
        long message_id = extras.getLong("message_id");
        if (message_id != -1) {
            Log.d(TAG, "Find the message_id and switch to edit mode.");
            request_uri = NetworkUtils.get_url(bot_token, "editMessageText");
            request_body.message_id = message_id;
        }
        String result_status = "Unknown";
        switch (getResultCode()) {
            case Activity.RESULT_OK:
                result_status = context.getString(R.string.success);
                break;
            case SmsManager.RESULT_ERROR_GENERIC_FAILURE:
                result_status = context.getString(R.string.send_failed);
                break;
            case SmsManager.RESULT_ERROR_RADIO_OFF:
                result_status = context.getString(R.string.airplane_mode);
                break;
            case SmsManager.RESULT_ERROR_NO_SERVICE:
                result_status = context.getString(R.string.no_network);
                break;
        }
        request_body.text = extras.getString("message_text") + "\n" + context.getString(R.string.status) + result_status;
        String request_body_raw = new Gson().toJson(request_body);
        RequestBody body = RequestBody.create(request_body_raw, const_value.JSON);
        OkHttpClient okhttp_client = NetworkUtils.get_okhttp_obj(sharedPreferences.getBoolean("doh_switch", true), Paper.book("system_config").read("proxy_config", new ProxyConfigV2()));
        Request request = new Request.Builder().url(request_uri).method("POST", body).build();
        Call call = okhttp_client.newCall(request);
        final String error_head = "Send SMS status failed:";
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                e.printStackTrace();
                LogUtils.write_log(context, error_head + e.getMessage());
                SmsUtils.send_fallback_sms(context, request_body.text, sub);
                ResendUtils.add_resend_loop(context, request_body.text);
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                if (response.code() != 200) {
                    assert response.body() != null;
                    LogUtils.write_log(context, error_head + response.code() + " " + Objects.requireNonNull(response.body()).string());
                    ResendUtils.add_resend_loop(context, request_body.text);
                }
            }
        });
    }
}

File: app/src/main/java/com/airfreshener/telegram_sms/receivers/SmsSendReceiver.kt
package com.airfreshener.telegram_sms.receivers

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.SmsManager
import android.util.Log
import com.airfreshener.telegram_sms.R
import com.airfreshener.telegram_sms.utils.ContextUtils.app
import com.airfreshener.telegram_sms.utils.ResendUtils
import com.airfreshener.telegram_sms.utils.SmsUtils

class SmsSendReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val app = context.app()
        val prefsRepository = app.prefsRepository
        val telegramRepository = app.telegramRepository
        Log.d(TAG, "Receive action: " + intent.action)
        context.applicationContext.unregisterReceiver(this)
        val extras = intent.extras
        if (extras == null) {
            Log.i(TAG, "Received extras is null")
            return
        }
        val sub = extras.getInt("sub_id")
        if (!prefsRepository.getInitialized()) {
            Log.i(TAG, "Uninitialized, SMS send receiver is deactivated.")
            return
        }
        val resultStatus = when (resultCode) {
            Activity.RESULT_OK -> context.getString(R.string.success)
            SmsManager.RESULT_ERROR_GENERIC_FAILURE -> context.getString(R.string.send_failed)
            SmsManager.RESULT_ERROR_RADIO_OFF -> context.getString(R.string.airplane_mode)
            SmsManager.RESULT_ERROR_NO_SERVICE -> context.getString(R.string.no_network)
            else -> "Unknown"
        }
        val message = extras.getString("message_text") + "\n" +
                context.getString(R.string.status) + resultStatus
        telegramRepository.sendMessage(
            message = message,
            messageId = extras.getLong("message_id"),
            onFailure = {
                SmsUtils.sendFallbackSms(context, message, sub)
                ResendUtils.addResendLoop(context, message)
            }
        )
    }

    companion object {
        private const val TAG = "SmsSendReceiver"
    }
}
